theon/src/main.rs

```diff
@@ -13,21 +13,244 @@
+//!    transferred to the supervisor.  For example, all
+//!    processors will be parked in the supervisor, running in a
+//!    fully-formed address space, etc.
+//! 2. Loading and initializing all of the binaries that make
+//!    up the system.  That is, the binary images of all
+//!    segments and tasks will be loaded, and segment
+//!    initializers will be invoked.
+//!
+//! Any failure at any point in this process is fatal, and boot
+//! fails.  Hence, we use the infallible allocation interface
+//! and make liberal use of panic!(), assert!(), and similar
+//! interfaces.
+//!
+//! ## Binary loading
+//!
+//! Binary images are loaded into memory by an earlier stage
+//! loader, and assumed to be resident once theon begins
+//! execution.  Theon will locate them, and load them into
+//! physical memory.
+//!
+//! Each binary is allocated a 16MiB region of physical RAM for
+//! its various pages; these regions begin at 64MiB and are
+//! aligned on 32MiB boundaries, giving us room for loading new
+//! images into the second 16MiBs of each binary's region for
+//! hitless update.
+//!
+//! Binaries represent either tasks or segments; see HDP 0002
+//! for a high level overview of the distinction.  For segments,
+//! the segment initializer will be invoked.  For tasks, the
+//! task will be loaded and an image prototype transferred into
+//! supervisor.
+//!
+//! ## Control Transfer
+//!
+//! Once Theon has initialized the system and loaded binaries,
+//! it will transfer control into the supervisor segment,
+//! passing a pointer to a serialized description of the system
+//! and its state.  This is the same system description that is
+//! given to the same entry path on hitless upgrade; thus, in
+//! some ways one may think of Theon as synthesizing a hitless
+//! update that starts the system running with no VMs.
+//!
+//! After theon has finished executing and transferred control
+//! into the supervisor, it will not run again, and it's
+//! resources --- in particular all memory associated with it,
+//! including its image --- are reclaimed.
+//!
+//! ## Colophon
+//!
+//! In antiquity, Theon was Hypatia's father.
+
 extern crate alloc;
 
 mod allocator;
+mod theon;
 mod x86_64;
 
+use alloc::vec::Vec;
+use core::ops::Range;
+
+use crate::x86_64::memory::{Region, Type};
+use arch::{VPageAddr, HPA, MIB, PF4K, V4KA};
+
+type Result<T> = core::result::Result<T, &'static str>;
+
+// Describes whether a given binary is a segment or a task,
+// see HDPs 0002, 0009, and 0010 for details.
+#[derive(Clone, Copy, Debug)]
+enum BinaryType {
+    Segment,
+    Task,
+}
+
+/// Metadata used in the binary table: the name of the binary,
+/// it's physical load address, and its type (either a segment
+/// or a task).
+type BinaryMeta = (&'static str, HPA, BinaryType);
+
+/// Binaries are loaded in 16MiB regions of physical memory
+/// that are aligned on 32MiB boundaries, starting at 64MiB.
+///
+/// This layout
```

Comment:
Incomplete sentence?

Reply:
Done.

(Removed; subsumed by the description in the file-level comment)